File: LeetCode/Kai/148SortList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode sortList(ListNode head) {
        if(head==null || head.next == null) return head;
        ListNode slow = head;
        ListNode fast = head;
        while(fast.next!=null && fast.next.next!=null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode mid = slow.next;
        slow.next = null;
        ListNode p1 = sortList(head);
        ListNode p2 = sortList(mid);
        return merge(p1, p2);   //divide and conquer, merge sort
    }
    
    public ListNode merge(ListNode p1, ListNode p2) {
        if(p1.val>p2.val) {
            ListNode temp = p2;
            p2 = p1;
            p1 = temp;
        }
        ListNode head = p1;
        while(p1!=null && p2!=null) {
            if(p1.next == null) {
                p1.next = p2;
                break;
            }
            if(p1.val<=p2.val && p1.next.val>=p2.val) {
                ListNode tmp = p2.next;
                p2.next = p1.next;
                p1.next = p2;
                p2 = tmp;
                p1 = p1.next;
            } else {
                p1 = p1.next;
            }
        }
        return head;
        
    }
    
}

File: LeetCode/Peng/148SortList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* findMiddle(ListNode* head) {
        ListNode* slow = head, *fast = head;
        while (fast && fast->next && fast->next->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }
    
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* dummyHead = new ListNode(-1);
        ListNode* current = dummyHead;
        while (l1 && l2) {
            if (l1->val < l2->val) {
                current->next = l1;
                l1 = l1->next;
            } else {
                current->next = l2;
                l2 = l2->next;
            }
            current = current->next;
        }
        if (l1) {
            current->next = l1;
        } else if (l2) {
            current->next = l2;
        }
        return dummyHead->next;
    }
    
    ListNode* sortList(ListNode* head) {
        ListNode* middle = findMiddle(head);
        if (!middle || !middle->next) {
            return head;
        }
        ListNode* next = middle->next;
        middle->next = NULL;
        return mergeTwoLists(sortList(head), sortList(next));
    }
};